### PR TITLE
Handle 128 bits trace Ids

### LINKED ISSUE #86: Support 128bit trace IDs
128bit trace IDs allow interoperability with systems like Google's tracing service, and less collision for those doing high-volume sampling (ex from devices or lambda functions).

The process includes adding a separate field TraceIdHigh which includes the upper 64 bits of the trace ID, or 0 if it 128-bit trace IDs aren't in use. Most implementations keep their sampling 64-bit even when 128-bit is on. A flag like below would be used to indicate whether or not zipkin4net would originate 128bit trace IDs or not. This would default to false to allow sites to slowly upgrade. In some sites it wouldn't be set to true ever.
```
    /** When true, new root spans will have 128-bit trace IDs. Defaults to false (64-bit) */
    public Builder traceId128Bit(boolean traceId128Bit) {
```

Key unit tests would include ensuring B3 propagation can accept the longer 32character trace IDs and not lose them on the way out (when applying to outbound requests). Other tests would show that the data sent to zipkin has traceIdHigh (if thrift) or a longer traceId field (if json)

https://github.com/openzipkin/b3-propagation/issues/6

### PR DESCRIPTION
**Pull request by fedj**
Fixes #86

### CODE REVIEW

**Comment by fedj** [comment]
cc @adriancole

File: zipkin4net/Criteo.Profiling.Tracing.UTest/Dispatchers/T_AsyncDispatcher.cs [hunk 1]
```diff
@@ -76,9 +76,10 @@ public void RecordShouldnotBeDispatchedIfStopped()
         public void RecordsShouldBeDispatchedInOrder()
         {
             var sync = new CountdownEvent(2);
-
-            var firstRecord = new Record(new SpanState(1, 0, 1, SpanFlags.None), TimeUtils.UtcNow, Annotations.ClientRecv());
-            var secondRecord = new Record(new SpanState(1, 0, 1, SpanFlags.None), TimeUtils.UtcNow, Annotations.ClientRecv());
+            var traceId = 1;
+            var traceIdHigh = 2;
+            var firstRecord = new Record(new SpanState(traceId, traceIdHigh, 0, 1, SpanFlags.None), TimeUtils.UtcNow, Annotations.ClientRecv());
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing.UTest/Dispatchers/T_AsyncDispatcher.cs]: nit I usually try to order traceIdHigh first, as it aligns with the hex encoding of traceIdHigh+traceId

File: zipkin4net/Criteo.Profiling.Tracing.UTest/T_SpanState.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿using NUnit.Framework;
+
+namespace Criteo.Profiling.Tracing.UTest
+{
+    [TestFixture]
+    public class T_SpanState
+    {
+        private const long traceId = 1;
+        private const long traceIdHigh = 0;
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing.UTest/T_SpanState.cs]: same nit on ordering, won't repeat again. feel free to go your way and I understand lexicographically this looks right

File: zipkin4net/Criteo.Profiling.Tracing.UTest/T_Trace.cs [hunk 3]
```diff
@@ -78,7 +79,7 @@ public void TraceCreatesCorrectRecordWithTimeSpecified()
         [TestCase(SpanFlags.SamplingKnown | SpanFlags.Sampled)]
         public void TraceSamplingForced(SpanFlags initialFlags)
         {
-            var spanState = new SpanState(1, 0, 1, initialFlags);
+            var spanState = new SpanState(1, 2, 0, 1, initialFlags);
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing.UTest/T_Trace.cs]: does csharp support constructor overloading? if so default traceIdHigh to zero

File: zipkin4net/Criteo.Profiling.Tracing.UTest/Transport/T_ZipkinHttpTraceInjectorExtractor.cs [hunk 4]
```diff
@@ -29,6 +31,18 @@ public void GetTraceThenSetHeadersEqualsOriginal(string encodedTraceId, string e
             Assert.AreEqual(flagsStr, recreatedHeaders[ZipkinHttpHeaders.Flags]);
         }
 
+        [Test]
+        public void Supports128BitsTraceId()
+        {
+            var traceId = "00FF0000000000001000000000002111";
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing.UTest/Transport/T_ZipkinHttpTraceInjectorExtractor.cs]: make a test that makes it obvious that the 00FF.. part is traceIdHigh? Ex by using shorter numbers like traceIdHigh=1, traceId=2 (as the hex is equiv to decimal and easy to see)

File: zipkin4net/Criteo.Profiling.Tracing/SpanState.cs [hunk 5]
```diff
@@ -27,9 +29,10 @@ public SamplingStatus SamplingStatus
         /// </summary>
         public SpanFlags Flags { get; private set; }
 
-        public SpanState(long traceId, long? parentSpanId, long spanId, SpanFlags flags)
+        public SpanState(long traceId, long traceIdHigh, long? parentSpanId, long spanId, SpanFlags flags)
         {
             TraceId = traceId;
+            TraceIdHigh = traceIdHigh;
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing/SpanState.cs]: default to zero if possible

File: zipkin4net/Criteo.Profiling.Tracing/SpanState.cs [hunk 6]
```diff
@@ -63,14 +69,15 @@ public override int GetHashCode()
...
+                hashCode = (hashCode * 397) ^ TraceIdHigh.GetHashCode();
                 hashCode = (hashCode * 397) ^ SpanId.GetHashCode();
                 return hashCode;
             }
         }
 
         public override string ToString()
         {
-            return string.Format("{0}.{1}<:{2}", TraceId, SpanId, ParentSpanId.HasValue ? ParentSpanId.Value.ToString(CultureInfo.InvariantCulture) : "_");
+            return string.Format("{0}{1}.{2}<:{3}", TraceIdHigh == 0 ? "" : TraceIdHigh.ToString(), TraceId, SpanId, ParentSpanId.HasValue ? ParentSpanId.Value.ToString(CultureInfo.InvariantCulture) : "_");
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing/SpanState.cs]: nice

File: zipkin4net/Criteo.Profiling.Tracing/Transport/ZipkinHttpTraceExtractor.cs [hunk 7]
```diff
@@ -76,5 +79,32 @@ public static bool TryParseTrace(string encodedTraceId, string encodedSpanId, st
...
+            }
+            var traceIdHighLength = encodedTraceId.Length - traceId64BitsSerializationLength;
+            var traceIdHighStr = encodedTraceId.Substring(0, traceIdHighLength);
+            return NumberUtils.DecodeHexString(traceIdHighStr);
+        }
+
+		/// <summary>
+		/// Extracts traceId. Detects if present and then decode the last 16 bytes
+		/// </summary>
+		private static long ExtractTraceId(string encodedTraceId)
```
Comment (codefromthecrypt) [inline zipkin4net/Criteo.Profiling.Tracing/Transport/ZipkinHttpTraceExtractor.cs]: nit indent

**Review by codefromthecrypt — APPROVED** [review]
LG only small things

**Comment by fedj** [comment]
@adriancole Done, I think it's easier to read with your comments. I also introduced a constant to see the absence of traceIdHigh

**Review by codefromthecrypt — APPROVED** [review]

**Review by crntn — APPROVED** [review]
I note an inconsistency between SpanState's Equals and GetHashCode that we can fix in another PR.
[https://github.com/criteo/zipkin4net/issues/91](url)

**Pull request merged**